keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -0,0 +1,1200 @@
+
+Once the connection is upgraded to a bi-directional streaming connection, the
+client and server can exchange data messages. These messages are interpreted with
+agreed upon standards which are called subprotocols. The three `kubectl` commands
+(`exec`, `attach`, and `cp`) communicate using the `RemoteCommand` subprotocol. Basically,
+this subprotocol provides command line functionality from the client to a running
+container in the cluster. By multiplexing `stdin`, `stdout`, `stderr`, and `tty`
+resizing over a streaming connection, this subprotocol supports clients executing
+and interacting with commands executed on a container in the cluster. An example of
+`kubectl exec` running the `date` command on an `nginx` pod/container is:
+
+```
+$ kubectl exec nginx -- date
+Tue May 16 03:34:04 PM PDT 2023
+```
+
+### Background: API Server and Kubelet `UpgradeAwareProxy`
+
+In order to route the data streamed between the client and the container, both the
+API Server and Kubelet must proxy these data messages. Both the API Server and the
+Kubelet provide this functionality with the `UpgradeAwareProxy`, which is a reverse
+proxy that knows how to deal with the connection upgrade handshake.
+
+### Proposal: `kubectl` WebSocket Executor and Fallback Executor
+
+This enhancement proposes adding a `WebSocketExecutor` to `kubectl`, implementing
+the WebSocket client using a new subprotol version (`v5.channel.k8s.io`). Additionally,
+we propose creating a `FallbackExecutor` to address client/server version skew. The
+`FallbackExecutor`  first attempts to upgrade the connection to WebSockets
+version five (`v5.channel.k8s.io`) with the `WebSocketExecutor`, then falls back
+to the legacy `SPDYExecutor` version four (`v4.channel.k8s.io`), if the upgrade is
+unsuccessful. Note that this mechanism can require two request/response trips instead
+of one. While the fallback mechanism may require an extra request/response if the
+initial upgrade is not successful, we believe this possible extra roundtrip is justified
+for the following reasons:
+
+1. The upgrade handshake is implemented in low-level SPDY and WebSocket libraries,
+and it is not easily exposed by these libraries. If it is even possible to modify
+the upgrade handshake, the added complexity would not be worth the effort.
+2. The streaming is already IO heavy, so another roundtrip will not substantially
+affect the perceived performance.
+3. As releases increment, the probablity of a WebSocket enabled `kubectl` communicating
+with an older non-WebSocket enabled API Server decreases.
+
+### Proposal: `v5.channel.k8s.io` subprotocol version
+
+As previously mentioned, we propose incrementing the subprotocol version from four
+to five: `v5.channel.k8s.io`. This version will indicate to upstream components
+that `kubectl` is interested in using WebSockets between the client and the API Server
+if it is supported.
+
+### Proposal: API Server `StreamTranslatorProxy`
```

Comment:
This is required to avoid problems with kubelets and runtimes. that doesn't support the v5.channel protocol that solves the gap in existing websocket implementation